### PR TITLE
[Merged by Bors] - simple tool to compare traces between executions

### PR DESCRIPTION
**Pull request by mockersf**
# Objective

- Have an easy way to compare spans between executions

## Solution

- Add a tool to compare spans from chrome traces

```bash
> cargo run --release  -p spancmp -- --help
   Compiling spancmp v0.1.0
    Finished release [optimized] target(s) in 1.10s
     Running `target/release/spancmp --help`
spancmp

USAGE:
    spancmp [OPTIONS] <TRACE> [SECOND_TRACE]

ARGS:
    <TRACE>
    <SECOND_TRACE>

OPTIONS:
    -h, --help                     Print help information
    -p, --pattern <PATTERN>        Filter spans by name matching the pattern
    -t, --threshold <THRESHOLD>    Filter spans that have an average shorther than the threshold
                                   [default: 0]
```

for each span, it will display the count, minimum duration, average duration and max duration. It can be filtered by a pattern on the span name or by a minimum average duration.

just displaying a trace
Screenshot 2022-04-28 at 21 56 21

comparing two traces
Screenshot 2022-04-28 at 21 56 55

### CODE REVIEW

**Comment by superdump** [comment]
1. Ooooooo
2. Aaaaahhh
3. ???
4. Profit!
5. Maybe a terminal UI and some more and selectable metrics and comparisons :)

**Comment by superdump** [comment]
Wait what? How did I miss that Tracy had that?? Awesome!

**Comment by alice-i-cecile** [comment]
bors try

**Comment by superdump** [comment]
> > This is something that tracy does too (most of it at least). This seems like a nice command line tool, but what is the argument for it to be included in bevy?
> 
> While I like tracy, it's an extra tool to install and more complicated to use.

Honestly, I agree. It's already clunky to switch branches to run the profiles. The Tracy workflow then involves clicking around in a GUI to open the traces, find the relevant systems in the two traces to compare, noting down the statistics to write into a discussion / issue / PR / whatever. If I could just capture the two profiles and then run a command that spits out the information I want, that's going to be a faster workflow.

> I added it in Bevy because it's not a general chrome trace comparison tool, it only supports the traces emitted by Bevy, and we can make it evolve based on the stats we want to monitor.

I'm in. And you know I'll make use of this so...

**Comment by alice-i-cecile** [comment]
I like the justification here. I'm on board with the design direction here, although I haven't reviewed the code in depth yet.

File: tools/spancmp/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+/// A span from the trace
+#[derive(Deserialize, Debug)]
+struct Span {
+    /// name
+    name: String,
+    /// phase
+    ph: String,
+    /// timestamp
+    ts: f32,
```
Comment (superdump) [inline tools/spancmp/src/main.rs:24]:
```suggestion
    name: String,
    phase: String,
    timestamp: f32,
```
and fix references. I see no reason to shorten the member names, nor to document with what the member names are/should be.

File: tools/spancmp/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    reader: BufReader<File>,
+    buf: Box<[u8; 1]>,
+    finish: String,
+}
+
+impl UnfinishedWrapper {
+    fn from(reader: BufReader<File>) -> UnfinishedWrapper {
+        Self {
+            reader,
+            // last: Ok(0),
```
Comment (superdump) [inline tools/spancmp/src/main.rs:53]: Left-over cruft?

File: tools/spancmp/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            buf: Box::new([0; 1]),
```
Comment (superdump) [inline tools/spancmp/src/main.rs:54]: Why Box? Is it needed?

File: tools/spancmp/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+    #[clap(short, long)]
+    /// Filter spans by name matching the pattern
+    pattern: Option<Regex>,
+
+    trace: String,
+    second_trace: Option<String>,
+}
+
+const MARGIN_PERCENT: f32 = 0.02;
```
Comment (superdump) [inline tools/spancmp/src/main.rs:88]: I just hacked this in. It should probably be a CLI argument with a default. :) Later we can base it on the standard deviation of the distributions or something statistically sensible.

File: tools/spancmp/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        } else {
+            Color::White
+        })))
+        .unwrap();
+    if v > 1.0 + MARGIN_PERCENT {
+        print!("+");
+    } else if v >= 1.0 - MARGIN_PERCENT {
+        print!(" ");
+    }
+    print!("{:5.2}%", if v.is_nan() { 1.0 } else { (v - 1.0) * 100.0 });
```
Comment (superdump) [inline tools/spancmp/src/main.rs:104]:
```suggestion
    print!("{:5.2}%", if v.is_nan() { 0.0 } else { (v - 1.0) * 100.0 });
```
My mistake here.

**Review by superdump — CHANGES_REQUESTED** [review]

File: tools/spancmp/src/main.rs (hunk 1 above)
Comment (mockersf) [reply tools/spancmp/src/main.rs:24]: The fields have the same name as in the json. They can be annotated with the serve annotation to rename them, but I tend to prefer to stick to the same name as the json being deserialised when not part of the public api

File: tools/spancmp/src/main.rs (hunk 3 above)
Comment (mockersf) [reply tools/spancmp/src/main.rs]: nop

**Comment by mockersf** [comment]
new look thanks to @superdump 
Screenshot 2022-05-04 at 01 36 18

**Comment by alice-i-cecile** [comment]
Can we use short names for the systems here? We already have a tool for this (get_short_name in bevy_reflect), that I refactored in https://github.com/bevyengine/bevy/pull/4299/files#diff-73324cf718e73f6475d5a1615002a1234d40afbb5f320ef0dd4d7aa337454866R1

**Comment by mockersf** [comment]
Almost, but not exactly: here the name is not directly a rust type name, it's a string, and we want to keep part of that string.

It's also useful to keep the full type to be able to filter all spans from `bevy_render` for example

**Comment by mockersf** [comment]
@alice-i-cecile finally, it's possible with some regex

I added a flag `--short` that will simplify system names. It has issues on some system names, we'll see once your updated version is merged if it's better
Screenshot 2022-05-04 at 03 07 22

**Comment by mockersf** [comment]
now with everything aligned and time units thanks to @superdump 
Screenshot 2022-05-04 at 21 06 20

**Review by superdump — APPROVED** [review]
I think we should merge this then continue iterating on it. :)

**Pull request closed without merging**